File: src/world/WorldEditTransaction.js
/**
 * Small synchronous transaction for cross-system world edits. Each step
 * registers its compensation before applying, so even a step that throws
 * after a partial mutation is included in reverse-order rollback.
 */
export class WorldEditTransaction {
  constructor(label = 'world-edit') {
    this.label = label;
    this.state = 'ACTIVE';
    this.compensations = [];
  }

  step(label, apply, compensate = () => {}) {
    if (this.state !== 'ACTIVE') throw new Error(`${this.label} transaction is ${this.state.toLowerCase()}`);
    if (typeof apply !== 'function' || typeof compensate !== 'function') {
      throw new TypeError('transaction steps require apply and compensate functions');
    }
    const entry = { label, compensate, value: undefined };
    this.compensations.push(entry);
    entry.value = apply();
    if (entry.value === false) {
      // `false` is the participant contract for a rejection without mutation.
      // Thrown errors retain the pre-registered compensation because a throw
      // may happen after a partial mutation.
      this.compensations.pop();
      throw new Error(`${label} rejected the ${this.label} transaction`);
    }
    return entry.value;
  }

  commit() {
    if (this.state !== 'ACTIVE') throw new Error(`${this.label} transaction is ${this.state.toLowerCase()}`);
    this.state = 'COMMITTED';
    this.compensations.length = 0;
    return true;
  }

  rollback(cause = null) {
    if (this.state !== 'ACTIVE') return [];
    const errors = [];
    for (let index = this.compensations.length - 1; index >= 0; index--) {
      const entry = this.compensations[index];
      try {
        entry.compensate(entry.value, cause);
      } catch (error) {
        errors.push({ step: entry.label, error });
      }
    }
    this.compensations.length = 0;
    this.state = 'ROLLED_BACK';
    return errors;
  }
}

export function runWorldEditTransaction(label, executor) {
  if (typeof executor !== 'function') throw new TypeError('transaction executor must be a function');
  const transaction = new WorldEditTransaction(label);
  try {
    const value = executor(transaction);
    transaction.commit();
    return value;
  } catch (error) {
    const rollbackErrors = transaction.rollback(error);
    if (rollbackErrors.length > 0) error.rollbackErrors = rollbackErrors;
    throw error;
  }
}

File: godot/MetroPulse.Domain/WorldEditing/WorldEditTransaction.cs
namespace MetroPulse.Domain.WorldEditing;

public enum WorldEditTransactionState
{
    Active,
    Committed,
    RolledBack,
}

public sealed record WorldEditRollbackError(string Step, Exception Error);

public sealed class WorldEditTransactionException : Exception
{
    public WorldEditTransactionException(
        string message,
        Exception? innerException = null,
        IReadOnlyList<WorldEditRollbackError>? rollbackErrors = null)
        : base(message, innerException)
    {
        RollbackErrors = rollbackErrors ?? Array.Empty<WorldEditRollbackError>();
    }

    public IReadOnlyList<WorldEditRollbackError> RollbackErrors { get; }
}

/// <summary>
/// Synchronous cross-authority transaction. Compensation is registered before apply,
/// and every applied or possibly partially-applied step rolls back in strict LIFO order.
/// </summary>
public sealed class WorldEditTransaction
{
    private sealed record Entry(string Label, Action<object?, Exception?> Compensate)
    {
        public object? Value { get; set; }
    }

    private readonly List<Entry> compensations = [];

    public WorldEditTransaction(string label)
    {
        if (string.IsNullOrWhiteSpace(label)) throw new ArgumentException("A transaction label is required.", nameof(label));
        Label = label.Trim();
    }

    public string Label { get; }

    public WorldEditTransactionState State { get; private set; } = WorldEditTransactionState.Active;

    public T Step<T>(string label, Func<T> apply, Action<T, Exception?>? compensate = null)
    {
        EnsureActive();
        if (string.IsNullOrWhiteSpace(label)) throw new ArgumentException("A step label is required.", nameof(label));
        ArgumentNullException.ThrowIfNull(apply);
        var entry = new Entry(label.Trim(), (value, cause) => compensate?.Invoke((T)value!, cause));
        entry.Value = default(T);
        compensations.Add(entry);
        T value = apply();
        if (value is bool accepted && !accepted)
        {
            compensations.RemoveAt(compensations.Count - 1);
            throw new WorldEditTransactionException($"{label} rejected the {Label} transaction.");
        }
        entry.Value = value;
        return value;
    }

    public void Commit()
    {
        EnsureActive();
        compensations.Clear();
        State = WorldEditTransactionState.Committed;
    }

    public IReadOnlyList<WorldEditRollbackError> Rollback(Exception? cause = null)
    {
        if (State != WorldEditTransactionState.Active) return Array.Empty<WorldEditRollbackError>();
        var errors = new List<WorldEditRollbackError>();
        for (int index = compensations.Count - 1; index >= 0; index--)
        {
            Entry entry = compensations[index];
            try
            {
                entry.Compensate(entry.Value, cause);
            }
            catch (Exception error)
            {
                errors.Add(new WorldEditRollbackError(entry.Label, error));
            }
        }
        compensations.Clear();
        State = WorldEditTransactionState.RolledBack;
        return errors.AsReadOnly();
    }

    public static T Run<T>(string label, Func<WorldEditTransaction, T> executor)
    {
        ArgumentNullException.ThrowIfNull(executor);
        var transaction = new WorldEditTransaction(label);
        try
        {
            T value = executor(transaction);
            transaction.Commit();
            return value;
        }
        catch (Exception error)
        {
            IReadOnlyList<WorldEditRollbackError> rollbackErrors = transaction.Rollback(error);
            if (error is WorldEditTransactionException transactionError && rollbackErrors.Count == 0)
            {
                throw transactionError;
            }
            throw new WorldEditTransactionException(
                $"{label} failed and was rolled back.",
                error,
                rollbackErrors);
        }
    }

    private void EnsureActive()
    {
        if (State != WorldEditTransactionState.Active)
        {
            throw new InvalidOperationException($"{Label} transaction is {State.ToString().ToLowerInvariant()}.");
        }
    }
}
